chapters/03-talking_to_servers/readme.md

```diff
@@ -0,0 +1,678 @@
+infer that it's for our Comment.
+
+To tap into this we need to change our form a little bit, instead of a
+`form_tag` we are going to use `form_for`. Instead of an URL it accepts an
+instance of a model and outputs a form builder,
+like so `<%= form_for Comment.new do |form_builder| %>`. Now our form is
+aware of the `Comment.new` context and all tags created with it will
+automatically be in the context of and bear attribute names of the `Comment`.
+E.g.:
+
+```html
+<%= form_for Comment.new do |f| %>
+  <%= f.text_area :content, size: '60x10' %>
+  <%= f.submit 'Post' %>
+<% end %>
+```
+
+If we now add this snippet at the bottom of the Post's show page, we can create
+comments from that page.
+
+```html
+<p id="notice"><%= notice %></p>
+
+<p>
+  <strong>Author:</strong>
+  <%= @post.author_id %>
+</p>
+
+<p>
+  <strong>Content:</strong>
+  <%= @post.content %>
+</p>
+
+<p>
+  <strong>Published:</strong>
+  <%= @post.published %>
+</p>
+
+<p>
+  <strong>Comments:</strong>
+</p>
+
+<p>
+  <br />
+  <strong>Post a comment:</strong>
+  <%= form_for Comment.new do |f| %>
+    <%= f.text_area :content, size: '60x10' %>
+    <%= f.submit 'Post' %>
+  <% end %>
+</p>
+
+<%= link_to 'Edit', edit_post_path(@post) %> |
+<%= link_to 'Back', posts_path %>
+```
+
+![Comment input with form_for](./images/comment_1.png)
+
+This is petty neat! If you click on post it will really create a comment
+and show it to you. Now we have the ability to create arbitrary inputs for our
+data types. But, as said, more on that in the next chapter.
+
+![Created Comment](./images/comment_2.png)
+
+Before we go on, we need to cover an even simpler method of creating forms for
+models - through the help of a Gem called `simple_form`. You've installed this
+gem in the previous chapter and therefore don't have to do it now.
+
+With simple form we get the best of both worlds. We can use form builders to
+be aware of the Model's context while still being able to just create HTML-like
+inputs and not care about the attribute type. In other words we can just use
+one tag `input` and it will infer the attribute type and input type form the
+model we passed to it. E.g. the above example can be written like this using
+simple_form:
+
+```html
+<%= simple_form_for Comment.new do |f| %>
+  <%= f.input :content %>
+  <%= f.button :submit %>
+<% end %>
+```
+
+![Comment input with simple_form](./images/comment_3.png)
+
+Notice that we didn't specify any type at all, yet if we render this view we
+will see that not only did we get a text area but everything looks much nicer
+than the previous solution. Simple form integrates with many CSS frameworks,
+such as Bootstrap, which we installed in the last chapter. From that it
+automatically applies everything necessary for our form to seamlessly integrate
+with Bootstrap.
+
+## Assignment
+
+1. Do all the steps in the chapter by your self.
+
+2. Create a new route, view and controller for a resource named `buildings`
```

Comment:
Can this assignment be changed so that students create something related to the global not-a-reddit task? For example students could create a login page without any complicated logic.

To put their newly acquired knowledge to the use, they could have 'captcha' like feature. Server asks them to input two numbers which solve a fixed equation. Eg. x + y - z = 11. On controller they should check the validity of the input and depending on that show error or redirect them home.

Reply:
1. No. We have a separate chapter for authentication. 
2. That would make sense if, in the end, the form was public, but it will be hidden behind a login screen. 

They have created a way to post comments on a post. In the next chapter we need to link this together with `belongs_to` and `has_many` and simple_form `collection` and `association`. But for now I don't see a way how to go into any depth as they don't have prior experiance with those concepts.

This chapter is intended to be a primer into client-server communication. There simply isn't anything they can do until we teach them how to work with the DB, and we can't teach them that first as then they can't visualize that without knowing how forms work first.

We can pull a lot together in the next chapter as now they know how to render, create resources and post data.

Reply:
I wouldn't implement any kind of authorization. Form would only check result of captcha. It wouldn't check if user exists of if the password is valid.

Reply:
But that would get deleted when we add Devise... I'm not really sure how to tie this in, so I opted not to.

Reply:
And what do we do with this solution and all previous ones like tic-tac-toe and employee-o-matic then delete them? It just seems more on topic. But for the sake of giving handout to students sooner, I'll approve this.

In future both of us need to minimize out-of-project examples and maybe rewrite them so that they are more closely related.

Reply:
> And what do we do with this solution and all previous ones like tic-tac-toe and employee-o-matic then delete them?

Yes, it was just a learning example. It is the Buddhist way - create and destroy, to create again.

>  In future both of us need to minimize out-of-project examples and maybe rewrite them so that they are more closely related.

If you are ok with that capcha later being deleted, then I'm ok with assigning it.